Removing excess check

toddbellMSFT opened the pull request:
We didn't need to add this check since there were underlying error checks to make sure things didn't go wrong. BUT we did find a bad #if defined() that should not exist.

Added checks to ensure if SessionTicket is set, DeveloperSecretKey isn't. and in the reverse too.

toddbellMSFT committed:
Removing excess check

We didn't need to add this check since there were underlying error checks to make sure things didn't go wrong. BUT we did find a bad #if defined() that should not exist.

Added checks to ensure if SessionTicket is set, DeveloperSecretKey isn't. and in the reverse too.

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -25,18 +25,18 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
 #ifndef DISABLE_PLAYFABENTITY_API
```
Gotta remove DISABLE_PLAYFABENTITY_API as well.

toddbellMSFT replied:
This is the wrong file. This is under 4.23/ExampleProject, not under private template.

pgilmorepf replied:
OH...
Yeah, undo all non-template changes.

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -25,18 +25,18 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-#endif
+        if (DeveloperSecretKey.Len() != 0)
+        {
+            checkf(false, TEXT("ClientSessionTicket is disabled because DISABLE_PLAYFABCLIENT_API is set."));
```
Change warning message.

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -92,35 +92,35 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-        checkf(false, TEXT("Cannot call GetDeveloperSecretKey in a non-developer build!"));
+        if (ClientSessionTicket.Len() != 0)
+        {
+            checkf(false, TEXT("Cannot call GetDeveloperSecretKey in a non-developer build!"));
```
Change warning message.

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -92,35 +92,35 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-#endif
+        if (ClientSessionTicket.Len() != 0)
+        {
+            checkf(false, TEXT("Cannot call GetDeveloperSecretKey in a non-developer build!"));
```
Change warning message...etc.

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -25,18 +25,18 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-#ifdef DISABLE_PLAYFABCLIENT_API
-        checkf(false, TEXT("ClientSessionTicket is disabled because DISABLE_PLAYFABCLIENT_API is set."));
-#endif
+        if (DeveloperSecretKey.Len() != 0)
```
Rethinking this...
We could collapse this to a single line:
checkf(DeveloperSecretKey.Len() == 0, TEXT("Cannot use ClientSessionTicket when DeveloperSecretKey is set."));

pgilmorepf commented on 4.23/ExampleProject/Plugins/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h:
```diff
@@ -92,35 +92,35 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
+        }
 
-        checkf(false, TEXT("Cannot call SetDeveloperSecretKey in a non-developer build!"));
+        DeveloperSecretKey = MoveTemp(InKey);
```
MoveTemp? Why not assign?
Not sure it's safe to do a move...

pgilmorepf commented on template/source/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h.ejs:
```diff
@@ -22,9 +22,24 @@ public:
 #ifndef DISABLE_PLAYFABENTITY_API
```
Must remove all instances of DISABLE_PLAYFABENTITY_API

pgilmorepf commented on template/source/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h.ejs:
```diff
@@ -22,9 +22,24 @@ public:
+            if(ClientSessionTicket.Len() != 0)
+            {
+                // TODO: Error here appropriately
+                checkf(false, TEXT("Cannot have a ClientSessionTicket with a Developer Secret Key Present!"));
```
```suggestion
                checkf(
                    DeveloperSecretKey.Len() != 0 && ClientSessionTicket.Len() != 0,
                    TEXT("Cannot have a ClientSessionTicket with a Developer Secret Key Present!")
                );
```

pgilmorepf reviewed (changes requested):
You missed some definitions to remove. Search for every "ABLE_PLAYFAB" in the Unreal template and remove them all.

An update from my earlier suggestion, checkf(bool, message) doesn't need to be wrapped in an if, because the first condition is the boolean to check.
Replace all instances of:
```if (condition) { checkf(false, message); }```
with
```checkf(!condition, message);```
And don't nest multiple ifs where relevant, just use:
```checkf(!condition1 || !condition2, message);```
Don't forget, the checkf condition is inverted from the if.

A lot of the warning messages are still referencing the old messages, about the defines, when we're actually changing it to the simultaneous use of secretKey and sessionTicket. Update the messages to the new condition.

pgilmorepf commented:
Upon further thought....
Every single one of those checks could be:
```checkf(DeveloperSecretKey.len() == 0 || ClientSessionTicket.len() == 0, "For title security, you cannot set the DeveloperSecretKey on a process which uses a Client Login");```
That's the right message no matter which is being set.
It checks both every time. It'll catch whenever the second one is set, every time.
It's easier to apply appropriately to every set/get for both variables.
No mental logic to carefully check. It always checks everything, and that's perfect.

toddbellMSFT committed:
Swap check function for a singular one.

pgilmorepf commented on template/source/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h.ejs:
```diff
@@ -16,109 +16,68 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-                checkf(false, TEXT("Cannot have a DeveloperSecretKey with a ClientSessionTicket Present!"));
-            }
-        }
+        checkf(DeveloperSecretKey.Len() != 0 && ClientSessionTicket.Len() != 0, TEXT("Cannot have a ClientSessionTicket with a Developer Secret Key Present!"));
```
```suggestion
        checkf(DeveloperSecretKey.len() == 0 || ClientSessionTicket.len() == 0, TEXT("Cannot have a ClientSessionTicket with a Developer Secret Key Present!"));
```

pgilmorepf replied:
Condition inverted here, probably because I had it wrong in my first suggestion (which I corrected in a later suggestion).
Also...
The difference between "Len" and "len" below scares me.
Make sure to thoroughly test this.

toddbellMSFT replied:
FString uses .Len() I'll make sure that is aligned.

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h

reverting 4.23 AuthenticationContext.

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h

missing last line

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

updating capitalization.

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

updating equality check

toddbellMSFT committed:
Merge branch 'master' into toddbell/dev-secret-key-fix

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

needed to surround string with TEXT()

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

fcheck needs the inverse of this check

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

double checking some logic. This should read as:

If d is something AND c is something

But we have to negate it, because passing false is what will stop the code running in the fcheck. We just made it true.

 SO The distirbution property should be applicable so seeing

d == 0 || c == 0

Should still logically also make sense (if both are 0...? then we have what we wanted to avoid)

idk... reading the last one sounds like "if d.length is 0 ||..." that first part is true, so it sounds like it would just skip the second part of the if check if the compiler was optimizing appropriately (it shouldn't need both sides of OR to be evaluated to true if the first argument evaluates to true).

THIS IS WHERE XOR WOULD BE HELPFUL (but that would be rude to have to force developers to read, this logically makes sense and forces the compiler to read how the engineers read. This is more maintainable then the xor).

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

making it more readable. Also updating the logic, it still doesn't like packaging (but runs fine in editor?)

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

Doesn't seem to compile unless this function is callable through blueprints?!

toddbellMSFT committed:
Update PlayFabAuthenticationContext.h.ejs

toddbellMSFT committed:
Update PlayFabTestRunner.cpp

The presence of both developerSecretKey and tests that login will fail.

We do not have any tests that require this dev secret key according to a local run on 4.24.

toddbellMSFT committed:
This is what's required

pgilmorepf commented on template/source/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h.ejs:
```diff
@@ -16,108 +16,75 @@ class PLAYFABCOMMON_API UPlayFabAuthenticationContext : public UObject
-#ifdef DISABLE_PLAYFABCLIENT_API
-        checkf(false, TEXT("ClientSessionTicket is disabled because DISABLE_PLAYFABCLIENT_API is set."));
-#endif
         ClientSessionTicket = InTicket;
```
```suggestion
        ClientAdminSecurityCheck();
        ClientSessionTicket = InTicket;
```

toddbellMSFT committed:
Update template/source/PlayFab/Source/PlayFabCommon/Public/PlayFabAuthenticationContext.h.ejs

Co-authored-by: Paul Gilmore <pagilmor@microsoft.com>

elitenas commented:
Hi,

This is causing crashes in the editor @pgilmorepf. Please see this - https://community.playfab.com/questions/47009/unreal-engine-client-log-in-crashunreal-engine-cra.html.

When should set the secret key?

elitenas commented:
@pgilmorepf @toddbellMSFT any update on this?
